video/out/wayland_common.c

```diff
@@ -796,6 +800,104 @@ static const struct wl_surface_listener surface_listener = {
+    // one more drawback the wayland version has. It is additionally assumes
+    // that everything the mpv window renders is something you actually care
+    // about timing. In other words, things like lua scripts that do not
+    // neccesarily draw in-step with the frames mpv renders will also be taken
+    // into consideration. There is no way for the compositor to know the
+    // difference so it will update its internal statistics whenever something
+    // new is on the mpv window. In other words, timings will probably be very,
+    // very wrong if scripts (or something else) constantly render onto the mpv
+    // window while video is playing. The most obvious example would be stats.lua.
+    if (sbc_passed) {
+        struct timespec ts;
+        if (clock_gettime(CLOCK_MONOTONIC, &ts)) {
+            return;
+        }
+
+        uint64_t now_monotonic = ts.tv_sec * 1000000LL + ts.tv_nsec / 1000;
+        uint64_t ust_mp_time = mp_time_us() - (now_monotonic - ust);
+        wl->last_sbc_mp_time = ust_mp_time;
+    }
+
+    if (!sbc)
+        return;
+
+    wl->last_queue_display_time = wl->last_sbc_mp_time + sbc_passed*wl->vsync_duration;
+}
+
+static void pres_set_clockid(void *data, struct wp_presentation *pres,
+                           uint32_t clockid)
+{
+    struct vo_wayland_state *wl = data;
+    wl->presentation = pres;
+    clockid = CLOCK_MONOTONIC;
+}
+
+static const struct wp_presentation_listener pres_listener = {
+    pres_set_clockid,
+};
+
+static void feedback_sync_output(void *data, struct wp_presentation_feedback *fback,
+                               struct wl_output *output)
+{
+}
+
+static void feedback_presented(void *data, struct wp_presentation_feedback *fback,
+                              uint32_t tv_sec_hi, uint32_t tv_sec_lo,
+                              uint32_t tv_nsec, uint32_t refresh_nsec,
+                              uint32_t seq_hi, uint32_t seq_lo,
+                              uint32_t flags)
+{
+    struct vo_wayland_state *wl = data;
+
+    // Very similar to oml_sync_control, in this case we assume that every
+    // time the compositor receives feedback, a buffer swap has been performed
+    // (which is a fair assumption according to the presentation time spec).
+    //
+    // Notes:
+    //  - tv_sec_lo + tv_sec_hi is the equivalent of oml's ust
+    //  - seq_lo + seq_hi is the equivalent of oml's msc
+    //  - these values are updated everytime the compositor receives feedback.
+
+    int64_t sec = (uint64_t) tv_sec_lo + ((uint64_t) tv_sec_hi << 32);
+    wl->ust = sec * 1000000LL + (uint64_t) tv_nsec / 1000;
+    wl->msc = (uint64_t) seq_lo + ((uint64_t) seq_hi << 32);
+
+    // save the ust/msc values of the last presented frame
+    wl->last_ust = wl->ust;
+    wl->last_msc = wl->msc;
+
+    // save the sbc and increment the sbc count when a frame is presented
+    wl->last_sbc = wl->sbc;
+    wl->sbc += 1;
```

Comment:
You can also use a queue instead of using heap objects.

Reply:
I don't think the heap object case applies here. In Wayland world, the frame callback tells the compositor when the next frame should actually be displayed. If multiple frames are created and waiting in line on the queue, it shouldn't matter because each one has to go through the frame callback (and the presentation callback). This approach guarantees each frame is operated on (sync_swap, buffer_swap, and everything). Basically the frame callback + presentation listener act as a sort of "blocking" pipeline that every frame must go through.

I'm probably updating `last_queue_display_time` at a less-than-optimal place though. I should do it with the eglswapbuffer call like with x11egl. Also, I don't think that sbc increment there is necessary either once I move the sync_swap.